File: sections/40_data_structures_mixed_contest/problems/l_dynamic_inversion_swaps/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n, q;
  if (!(cin >> n >> q))
    return 0;
  vector<int> a(n);
  for (int &x : a)
    cin >> x;
  vector<int> bit(n + 1);
  auto add = [&](int i) {
    for (; i <= n; i += i & -i)
      bit[i]++;
  };
  auto sum = [&](int i) {
    int s = 0;
    for (; i; i -= i & -i)
      s += bit[i];
    return s;
  };
  ll inv = 0;
  for (int x : a) {
    inv += sum(n) - sum(x);
    add(x);
  }
  int B = 450, nb = (n + B - 1) / B;
  vector<vector<int>> blk(nb);
  auto rebuild = [&](int b) {
    blk[b].assign(a.begin() + b * B, a.begin() + min(n, (b + 1) * B));
    sort(blk[b].begin(), blk[b].end());
  };
  for (int b = 0; b < nb; b++)
    rebuild(b);
  auto count = [&](int l, int r, int lo, int hi) {
    int ans = 0;
    if (l > r || lo >= hi)
      return 0;
    while (l <= r && l % B) {
      ans += lo < a[l] && a[l] < hi;
      l++;
    }
    while (l + B - 1 <= r) {
      auto &v = blk[l / B];
      ans += lower_bound(v.begin(), v.end(), hi) -
             upper_bound(v.begin(), v.end(), lo);
      l += B;
    }
    while (l <= r) {
      ans += lo < a[l] && a[l] < hi;
      l++;
    }
    return ans;
  };
  while (q--) {
    int i, j;
    cin >> i >> j;
    --i;
    --j;
    if (i > j)
      swap(i, j);
    if (i != j) {
      int x = a[i], y = a[j], c = count(i + 1, j - 1, min(x, y), max(x, y));
      ll d = 1 + 2LL * c;
      inv += (x < y ? d : -d);
      swap(a[i], a[j]);
      rebuild(i / B);
      if (i / B != j / B)
        rebuild(j / B);
    }
    cout << inv << '\n';
  }
}

File: sections/40_data_structures_mixed_contest/problems/l_dynamic_inversion_swaps/solution.py
import bisect,math,sys
def main():
    d=list(map(int,sys.stdin.buffer.read().split()))
    if not d:return
    n,q=d[:2];a=d[2:2+n];bit=[0]*(n+1);inv=0
    def pref(i):
        s=0
        while i:s+=bit[i];i-=i&-i
        return s
    for seen,x in enumerate(a):
        inv+=seen-pref(x)
        i=x
        while i<=n:bit[i]+=1;i+=i&-i
    B=max(1,int(math.sqrt(n))+1);nb=(n+B-1)//B;blocks=[]
    for b in range(nb):blocks.append(sorted(a[b*B:min(n,(b+1)*B)]))
    def rebuild(b):blocks[b]=sorted(a[b*B:min(n,(b+1)*B)])
    def count(l,r,lo,hi):
        ans=0
        while l<=r and l%B:ans+=lo<a[l]<hi;l+=1
        while l+B-1<=r:
            v=blocks[l//B];ans+=bisect.bisect_left(v,hi)-bisect.bisect_right(v,lo);l+=B
        while l<=r:ans+=lo<a[l]<hi;l+=1
        return ans
    out=[];p=2+n
    for _ in range(q):
        i,j=d[p]-1,d[p+1]-1;p+=2
        if i>j:i,j=j,i
        if i!=j:
            x,y=a[i],a[j];delta=1+2*count(i+1,j-1,min(x,y),max(x,y));inv+=delta if x<y else -delta
            a[i],a[j]=y,x;rebuild(i//B)
            if i//B!=j//B:rebuild(j//B)
        out.append(str(inv))
    print('\n'.join(out))
if __name__=="__main__":main()
